### PR TITLE
Msf profile

### PR DESCRIPTION
**Pull request by SAkshat**
First step of multipart signup form, Add Profile.

### CODE REVIEW

File: app/controllers/agent/profile_controller.rb [hunk 1]
```diff
@@ -1,4 +1,31 @@
...
+  before_action :build_nested_associations, only: [:show]
+
+  def show
+    render_wizard
+  end
+
+  def update
+    current_agent.update(agent_params)
+    current_agent.state = step
+    render_wizard current_agent
```
Comment (ShubhamGupta) [inline app/controllers/agent/profile_controller.rb]: About convention - Use parenthesis around arguments while calling methods. However, you can skip them in declaration. But, follow one convention throughout the codebase.

File: app/models/address.rb [hunk 2]
```diff
@@ -1,6 +1,8 @@
 class Address < ActiveRecord::Base
   belongs_to :state
   belongs_to :city
-
```
Comment (ShubhamGupta) [inline app/models/address.rb]: Add a validation -- `city` must belong to the `state`.

File: db/migrate/20151106061137_add_phone_and_contact_email_fields_to_agent.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class AddPhoneAndContactEmailFieldsToAgent < ActiveRecord::Migration
```
Comment (ShubhamGupta) [inline db/migrate/20151106061137_add_phone_and_contact_email_fields_to_agent.rb]: No need to change this one in particular but for future, as long as we can, lets not build new migrations. Edit the existing ones for each table.

File: app/views/agent/profile/add_personal_info.html.erb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+
+<%= simple_form_for current_agent, url: wizard_path do |f| %>
+  <%= f.input :first_name %>
+  <%= f.input :last_name %>
+  <%= f.input :license_number %>
+  <%= f.input :phone_number %>
+  <%= f.input :contact_email %>
+
+  <%= f.simple_fields_for :address do |address_f| %>
```
Comment (ShubhamGupta) [inline app/views/agent/profile/add_personal_info.html.erb]: Lets not use `address_f`. Either use `address_fields` or just `h`. `h` is a well recognised convention.

File: app/views/agent/profile/add_personal_info.html.erb [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    <%= address_f.input :line1 %>
+    <%= address_f.input :state_id, collection: State.all, label_method: :name, value_method: :id, label: :State, include_blank: false %>
+    <%= address_f.input :city_id, collection: City.all, label_method: :name, value_method: :id, label: :City, include_blank: false %>
```
Comment (ShubhamGupta) [inline app/views/agent/profile/add_personal_info.html.erb]: What is the difference b/w `label_method` and `label`?

File: app/validators/email_uniqueness_validator.rb [hunk 6]
```diff
@@ -3,7 +3,10 @@ class EmailUniquenessValidator < ActiveModel::EachValidator
   MODELS = [:Agent, :Business]
 
   def validate_each(record, attribute, value)
-    if MODELS.any? { |m| m.to_s.constantize.exists?(email: value) }
+    if MODELS.any? do |m|
+        object = m.to_s.constantize.find_by(email: value)
+        object != record if object
```
Comment (ShubhamGupta) [inline app/validators/email_uniqueness_validator.rb]: `object && object != record`

File: app/models/city.rb [hunk 7]
```diff
@@ -1,3 +1,5 @@
 class City < ActiveRecord::Base
   belongs_to :state
+
```
Comment (ShubhamGupta) [inline app/models/city.rb]: For these models like `city` / `state` / `services` etc do not allow them to be deleted. These are the core data that supports the app and are required for normal functioning of app.

File: app/models/agent.rb [hunk 8]
```diff
@@ -4,7 +4,25 @@ class Agent < ActiveRecord::Base
...
-  validates :email, presence: true, email_format: true, email_uniqueness: true
+  validates :email, email_format: true, email_uniqueness: true
+
+  with_options if: -> { in_state?(:add_personal_info) } do
+    validates :first_name, :last_name, :license_number, :email, :phone_number, :contact_email, presence: true
+    validates :phone_number, numericality: { only_integer: true }
+    validates :contact_email, email_format: true
+    validates_associated :address
+  end
+  validates :phone_number, phony_plausible: true if "in_state?(:add_personal_info)"
```
Comment (ShubhamGupta) [inline app/models/agent.rb]: Don't use strings. Create a separate private method.

**Comment by ShubhamGupta** [comment]
@SAkshat I have reviewed the PR. Please do the changes and assign it to @gauravmahajan31 for a final round of CR. Overall good work.

File: app/controllers/agent/profile_controller.rb [hunk 9]
```diff
@@ -1,4 +1,30 @@
-class Agent::ProfileController < Wicked::WizardController
+class Agent::ProfileController < Agent::BaseController
+
+  include Wicked::Wizard
+
+  steps :add_personal_info
+
+  before_action :build_nested_associations, only: [:show]
```
Comment (gmahajan31) [inline app/controllers/agent/profile_controller.rb]: Can this be moved to after callback?

File: app/views/agent/profile/add_personal_info.html.erb (hunk 5 above)
Comment (SAkshat) [inline app/views/agent/profile/add_personal_info.html.erb]: label is used to set a label for the input tag. label_method is used to change the visible names of the items in the select box similar to how label_value changes the value of the object.

File: app/models/policy.rb [hunk 10]
```diff
@@ -1,2 +1,3 @@
 class Policy < ActiveRecord::Base
```
Comment (ShubhamGupta) [inline app/models/policy.rb]: This model should be `policy_type`. Skipped in first round.

File: app/models/concerns/indestructible.rb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+module Indestructible
+
+  extend ActiveSupport::Concern
+
+  included do
+    before_destroy :stop_destruction
+
+    def stop_destruction
+      false
```
Comment (ShubhamGupta) [inline app/models/concerns/indestructible.rb]: Add an error message.

File: app/models/agent.rb [hunk 12]
```diff
@@ -4,7 +4,29 @@ class Agent < ActiveRecord::Base
...
+  end
+  validates :phone_number, phony_plausible: true if :in_personal_info_state?
+
+  private
+
+    def in_state?(step)
+      state && state.to_sym == step
+    end
+
+    def in_personal_info_state?
```
Comment (ShubhamGupta) [inline app/models/agent.rb]: Need to discuss this.

File: app/models/agent.rb [hunk 13]
```diff
@@ -4,7 +4,29 @@ class Agent < ActiveRecord::Base
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
```
Comment (ShubhamGupta) [inline app/models/agent.rb]: Please make a constant `REGISTRATION_STEPS`

**Pull request merged**